_source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md

```diff
@@ -0,0 +1,362 @@
+
+To validate access tokens remotely, update your `validate` function, and add the following `validate_remotely` function:
+
+```python
+
+def validate_remotely(token, issuer, clientId, clientSecret):
+    headers = {
+        'accept': 'application/json',
+        'cache-control': 'no-cache',
+        'content-type': 'application/x-www-form-urlencoded',
+    }
+    data = {
+        'client_id': clientId,
+        'client_secret': clientSecret,
+        'token': token,
+    }
+    url = issuer + '/v1/introspect'
+
+    response = httpx.post(url, headers=headers, data=data)
+
+    return response.status_code == httpx.codes.OK and response.json()['active']
+
+
+def validate(token: str = Depends(oauth2_scheme)):
+    res = validate_remotely(
+        token,
+        config('OKTA_ISSUER'),
+        config('OKTA_CLIENT_ID'),
+        config('OKTA_CLIENT_SECRET')
+    )
+
+    if res:
+        return True
+    else:
+        raise HTTPException(status_code=400)
+```
+
+Now, refresh the docs, generate an access token by entering your Okta client ID and secret again, and call the `/items` endpoint. You should see a `200 Success` response again with the items. This endpoint is now calling the Okta authorization server on every request.
+
+### Validating Access Tokens Locally
+
+Because access tokens are generally short-lived (an hour by default), you might [prefer to validate the tokens locally](https://developer.okta.com/docs/guides/validate-access-tokens/go/overview/). This method is slightly less secure because you can't be sure that the access token hasn't been revoked remotely, but on the other hand, you don't have to use your Okta client secret to validate the token locally. Finally, local validation is significantly faster because you don't have to make a request to the Okta authorization server with every API call. You'll have to decide which method is most appropriate for your application.
+
+To validate the access token locally, install the [Okta JWT Python package](https://github.com/adithyasampatoor/okta_jwt):
+
+```bash
+pip install okta_jwt
+pip freeze > requirements.txt
+```
+
+Next, import the package's `validate_token` function and update the `validate` function in your `main.py` file:
+
+```python
+...
+from okta_jwt.jwt import validate_token as validate_locally
+
+...
+
+def validate(token: str = Depends(oauth2_scheme)):
+    try:
+        res = validate_locally(
+            token,
+            config('OKTA_ISSUER'),
+            config('OKTA_AUDIENCE'),
+            config('OKTA_CLIENT_ID')
+        )
+        return bool(res)
+    except Exception:
+        raise HTTPException(status_code=403)
+...
+```
+
+When you call the `/items` endpoint, the API will decode the JWT and validate it locally. The decoded JWTs are cached, so subsequent requests will be faster than the first one.
+
+## Conclusion
```

Comment:
I'd change this to "Learn More About Python and REST APIs" so it's a better SEO header.